From: Tsutomu Itoh <t-itoh@jp.fujitsu.com>
Date: Thu, 1 Dec 2011 19:21:22 +0900
Patch-mainline: no
References: FATE#306586
Subject: [PATCH] Btrfs: forced readonly when orphan_del fails

In the case where the orphan_del fails, and cannot return
error to the caller, the filesystem turns to readonly instead
of BUG_ON().

Signed-off-by: Tsutomu Itoh <t-itoh@jp.fujitsu.com>
Signed-off-by: David Sterba <dsterba@suse.cz>
---
 fs/btrfs/inode.c |    8 ++++++--
 1 files changed, 6 insertions(+), 2 deletions(-)

diff --git a/fs/btrfs/inode.c b/fs/btrfs/inode.c
index fc888fd..1089482 100644
--- a/fs/btrfs/inode.c
+++ b/fs/btrfs/inode.c
@@ -1959,7 +1959,10 @@ void btrfs_orphan_commit_root(struct btrfs_trans_handle *trans,
 	    btrfs_root_refs(&root->root_item) > 0) {
 		ret = btrfs_del_orphan_item(trans, root->fs_info->tree_root,
 					    root->root_key.objectid);
-		BUG_ON(ret);
+		if (ret) {
+			btrfs_std_error(root->fs_info, ret);
+			return;
+		}
 		root->orphan_item_inserted = 0;
 	}
 
@@ -3521,7 +3524,8 @@ void btrfs_evict_inode(struct inode *inode)
 	if (ret == 0) {
 		trans->block_rsv = root->orphan_block_rsv;
 		ret = btrfs_orphan_del(trans, inode);
-		BUG_ON(ret);
+		if (ret)
+			btrfs_std_error(root->fs_info, ret);
 	}
 
 	trans->block_rsv = &root->fs_info->trans_block_rsv;
-- 
1.7.6.233.gd79bc

